### PR TITLE
add host-style buckets support

### LINKED ISSUE #208: [FEATURE] add support for virtual-host-style buckets
### Description
`chorus` currently parses S3 bucket/object in two modes:

- path-style (/bucket/object)
- virtual-host-style (bucket.<proxy-host>/object)

The virtual-host logic is handled in `pkg/s3/req_parser.go` via `ParseBucketAndObjectForHost`, while generic callers still use `ParseBucketAndObject(r)` with an empty host context.

### Problem or Use Case
When virtual-host-style is expected, requests can fall through as `UndefinedMethod` if host matching config is missing/misconfigured, e.g. PUT / for bucket creation (s3cmd mb s3://test ...) gets rejected with `NotImplemented`.

### Proposed Solution
With `proxy.virtual_hostname` parameter configured, host-style requests should be parsed and routed correctly:

bucket extracted from request host prefix
object extracted from request path
method resolved as normal S3 API (CreateBucket, etc.)

Conditions:
- Ensure all proxy request parsing paths use host-aware parser with configured virtual_hostname.
- Keep path-style behavior unchanged for non-host-style requests.

Add/extend tests for:
- PUT / host-style bucket create
- host-style object operations
- fallback to path-style parsing

### Backwards compatibility
No problems expected

### PR DESCRIPTION
**Pull request by askarsabyrov**
### Description

Resolves #208 
This PR adds support for S3 virtual-host-style requests in the proxy while keeping existing path-style behavior unchanged.

It introduces host-aware bucket/object parsing in the S3 request parser and wires proxy middleware/auth to use the configured proxy.virtual_hostname value. With this change, requests like PUT / on bucket.<virtual_hostname> are correctly recognized as bucket operations (e.g. `CreateBucket`) instead of falling through to `UndefinedMethod`.

Key updates:
- Add host-aware parsing path in `pkg/s3/req_parser.go` (ParseReqForHost, ParseBucketAndObjectForHost).
- Use host-aware parser in proxy S3 middleware.
- Update signature-v2 canonical resource handling for virtual-host-style requests.
- Rewrite forwarded requests to backend path-style when incoming request is virtual-host-style.
- Add virtual_hostname to proxy/standalone config wiring and pass it through startup paths.
- Add/adjust unit tests for parser, auth resource canonicalization, and forward-path rewriting.

### Checklist
- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) file.
- [x] My commits include a `Signed-off-by` line to certify agreement with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).
- [x] `make test` passing
- [x] I have updated documentation in [README.md](README.md) if applicable.

**Note**: By submitting this PR, you agree to license your contributions under the [Apache 2.0 License](LICENSE) and follow our [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: service/proxy/config.yaml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 address: "http://localhost:9669" # Chorus proxy s3 api address
+virtual_hostname: "" # optional hostname or URL suffix for virtual-host-style buckets, e.g. s3.example.com
```
Comment (arttor) [inline service/proxy/config.yaml]: all existing chorus config properties using camelCase.


```suggestion
virtualHostname: "" # optional hostname or URL suffix for virtual-host-style buckets, e.g. s3.example.com
```

File: pkg/config/loader.go [hunk 2]
```diff
@@ -91,28 +91,36 @@ func setMapValue(m map[string]any, path []string, value string) bool {
 		return false
 	}
 
-	var foundKey string
-	for k := range m {
-		if strings.EqualFold(k, path[0]) {
-			foundKey = k
-			break
+	for width := len(path); width >= 1; width-- {
```
Comment (arttor) [inline pkg/config/loader.go]: what was the motivation behind changing config package? how it is related to virtual hosts?

File: service/proxy/config_test.go [hunk 3]
```diff
@@ -70,4 +70,12 @@ auth:
 		r.NoError(err)
 		r.Equal("encryption-key!!", conf.Storage.DynamicCredentials.MasterPassword)
 	})
+
+	t.Run("CFG_VIRTUAL_HOSTNAME", func(t *testing.T) {
```
Comment (arttor) [inline service/proxy/config_test.go]: correct envar name should be "CFG_VIRTUALHOSTNAME". see: https://docs.clyso.com/docs/products/chorus/configuration/

File: service/proxy/config.go [hunk 4]
```diff
@@ -48,11 +48,12 @@ type Storage = objstore.GenericStorage[*s3.Storage, *router.SwiftStorage]
...
-	Storage Storages     `yaml:"storage,omitempty"`
-	Auth    *auth.Config `yaml:"auth,omitempty"`
-	Port    int          `yaml:"port"`
-	Address string       `yaml:"address"`
-	Cors    *cors.Config `yaml:"cors"`
+	Storage         Storages     `yaml:"storage,omitempty"`
+	Auth            *auth.Config `yaml:"auth,omitempty"`
+	Port            int          `yaml:"port"`
+	Address         string       `yaml:"address"`
+	VirtualHostname string       `yaml:"virtual_hostname"`
```
Comment (arttor) [inline service/proxy/config.go]: how `VirtualHostname` is different from chorus proxy `Address`? Can address be used to detect/parse host bucket name?

File: pkg/s3/req_parser_test.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+	tests := []struct {
+		name            string
+		target          string
+		host            string
+		endpointAddress string
+		wantBucket      string
+		wantObject      string
+		wantVirtualHost bool
+	}{
```
Comment (arttor) [inline pkg/s3/req_parser_test.go:21]: can you please add more negative test cases and edge cases, like: endpointAddress="" (passthrough), host == endpointHost exactly (root, not a bucket), suffix-mismatch host (wrong.example.com), Host: .s3.example.com (degenerate empty bucket).

File: service/proxy/auth/signature_v2_test.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		},
+		{
+			name:            "virtual host object",
+			path:            "/object/nested",
+			host:            "bucket.s3.example.com",
+			endpointAddress: "http://s3.example.com",
+			want:            "/bucket/object/nested",
+		},
+		{
+			name:            "virtual host bucket root",
```
Comment (arttor) [inline service/proxy/auth/signature_v2_test.go:39]: it it possible to add a test that signs a real SigV2 request with minio-go's SigV2 signer or other go sdk and asserts doesSignatureV2Match accepts it?

**Review by arttor — CHANGES_REQUESTED** [review]
Thanks for submitting the issue and PR!

can you please  rebase onto current main - otherwise CI check will not pass.

what do you think about the following approch:
- reuse `address` config for detecting host bucket
- detect bucket from r.Host and rewrite r.URL.Path to path-style once in `S3Middleware`. 
- auth middleware check runs before req-parse, so SigV4/SigV2 verify the original wire form; the rewrite is invisible to both and there is no need to pass Get/SetVirtualHostStyle to ctx. 

aloso, do you have any strategy for e2e testing? not sure that it will be easy to test it on CI with e2e test? have you tested your branch manually?

File: service/proxy/config.go (hunk 4 above)
Comment (askarsabyrov) [reply service/proxy/config.go]: I was accounting for cases where a single Ceph RGW hosts multiple hostnames, though I realize this is a bit of an edge case now.

Comment (askarsabyrov) [reply service/proxy/config.go]: I'll drop the virtual_hostname parameter at all

File: service/proxy/config.yaml (hunk 1 above)
Comment (askarsabyrov) [reply service/proxy/config.yaml]: No longer relevant

File: service/proxy/config_test.go (hunk 3 above)
Comment (askarsabyrov) [reply service/proxy/config_test.go]: No longer relevant also

**Comment by askarsabyrov** [comment]
Thanks for the feedback.

Changes made:
- rebased onto current main
- removed separate virtual_hostname / VirtualHostname config
- reused existing proxy address for virtual-host bucket detection
- reverted the config loader change, since it is no longer needed
- moved virtual-host path rewrite into S3Middleware, after auth and before routing/forwarding

All tests passed.

File: pkg/config/loader.go (hunk 2 above)
Comment (askarsabyrov) [reply pkg/config/loader.go]: no need

File: pkg/s3/req_parser_test.go (hunk 5 above)
Comment (askarsabyrov) [reply pkg/s3/req_parser_test.go:21]: added, take a look please if its enough

File: service/standalone/config.yaml [hunk 7]
```diff
@@ -19,6 +19,7 @@ lock:
 proxy:
   enabled: true
   address: http://localhost:9669 # Chorus proxy s3 api address
+  virtual_hostname: "" # optional hostname or URL suffix for virtual-host-style buckets, e.g. s3.example.com
```
Comment (arttor) [inline service/standalone/config.yaml]: should be removed.

**Review by arttor — CHANGES_REQUESTED** [review]
thanks for updates but there are some comments and questions not adressed from the previous review round.

File: service/standalone/config.yaml (hunk 7 above)
Comment (askarsabyrov) [reply service/standalone/config.yaml]: thanks, missed this one

File: service/proxy/auth/signature_v2_test.go (hunk 6 above)
Comment (askarsabyrov) [reply service/proxy/auth/signature_v2_test.go:39]: added function [TestDoesSignatureV2MatchAcceptsSignedVirtualHostRequest](https://github.com/clyso/chorus/pull/206/changes#diff-a40671d4101bfffb317ab7690b0f2903e22dd52149f3ef57c678f8117e7d67cbR58), which signs request with `github.com/minio/minio-go/v7/pkg/signer` lib and ensures signatures matches

**Comment by askarsabyrov** [comment]
@arttor, I've addressed all the comments and resolved the discussions. Please let me know if I missed anything.

**Comment by askarsabyrov** [comment]
Tested the branch locally
haproxy(in front of chorus) logs:
```
May 21 10:34:49  awesomebucket.archive.pscloud.io 10.216.232.112:57074 http_frontend chorus-proxy "HEAD /chorus.yaml HTTP/1.1" 200 458 33
May 21 10:34:50  awesomebucket.archive.pscloud.io 10.216.232.112:57074 http_frontend chorus-proxy "GET /chorus.yaml HTTP/1.1" 200 442 59
```
chorus debug logs:
```
proxy-1    | 2026-05-21T10:39:09Z DBG ../../build/service/proxy/router/server.go:125 > proxy: new request received app=proxy app_id=d87e1hmhpios738bij8g bucket=awesomebucket flow=event http_method=GET http_path=/awesomebucket/ http_query=location method=GetBucketLocation stor_type=S3 user=10390
proxy-1    | 2026-05-21T10:39:09Z DBG ../../build/pkg/ratelimit/service.go:120 > rate limit disabled app=proxy app_id=d87e1hmhpios738bij8g bucket=awesomebucket flow=event http_method=GET http_path=/awesomebucket/ http_query=location method=GetBucketLocation stor_name=object stor_type=S3 user=10390
proxy-1    | 2026-05-21T10:39:09Z DBG ../../build/service/proxy/router/server.go:125 > proxy: new request received app=proxy app_id=d87e1hmhpios738bij8g bucket=awesomebucket flow=event http_method=HEAD http_path=/chorus.yaml http_query= method=HeadObject obj_name=chorus.yaml stor_type=S3 user=10390
proxy-1    | 2026-05-21T10:39:09Z DBG ../../build/pkg/ratelimit/service.go:120 > rate limit disabled app=proxy app_id=d87e1hmhpios738bij8g bucket=awesomebucket flow=event http_method=HEAD http_path=/chorus.yaml http_query= method=HeadObject obj_name=chorus.yaml stor_name=object stor_type=S3 user=10390
```

File: service/standalone/config.go [hunk 8]
```diff
@@ -57,12 +57,13 @@ type Config struct {
...
-		Auth    *auth.Config   `yaml:"auth,omitempty"`
-		Port    int            `yaml:"port"`
-		Address string         `yaml:"address"`
-		Cors    *cors.Config   `yaml:"cors"`
+		Storage         proxy.Storages `yaml:"storage,omitempty"`
+		Enabled         bool           `yaml:"enabled"`
+		Auth            *auth.Config   `yaml:"auth,omitempty"`
+		Port            int            `yaml:"port"`
+		Address         string         `yaml:"address"`
+		VirtualHostname string         `yaml:"virtual_hostname"`
```
Comment (arttor) [inline service/standalone/config.go]: is it still used?

File: service/proxy/auth/signature_v2.go [hunk 9]
```diff
@@ -275,12 +275,17 @@ func unescapeQueries(encodedQuery string) (unescapedQueries []string, err error)
 }
 
 // Returns "/bucketName/objectName" for path-style or virtual-host-style requests.
-func getResource(path string, _ string, domains []string) (string, error) {
-	if len(domains) == 0 {
-		return path, nil
+func getResource(path string, host string, endpointAddress string) (string, error) {
+	if bucket, _, virtualHost := s3.ParseBucketAndObjectForHost(&http.Request{
```
Comment (arttor) [inline service/proxy/auth/signature_v2.go]: do we need full http request here? is it makes sense to reuse `bucketFromHost` from req_parser.go?

File: pkg/config/loader.go [hunk 10]
```diff
@@ -98,7 +98,6 @@ func setMapValue(m map[string]any, path []string, value string) bool {
 			break
 		}
 	}
-
```
Comment (arttor) [inline pkg/config/loader.go:101]: Unrelated whitespace-only deletion

File: service/proxy/config_test.go [hunk 11]
```diff
@@ -70,4 +70,5 @@ auth:
 		r.NoError(err)
 		r.Equal("encryption-key!!", conf.Storage.DynamicCredentials.MasterPassword)
 	})
+
```
Comment (arttor) [inline service/proxy/config_test.go]: Unrelated whitespace-only change. Please revert.

**Review by arttor — CHANGES_REQUESTED** [review]

File: service/standalone/config.go (hunk 8 above)
Comment (askarsabyrov) [reply service/standalone/config.go]: not used, sorry, i got messed changes when squashing commits, removed already

File: pkg/config/loader.go (hunk 10 above)
Comment (askarsabyrov) [reply pkg/config/loader.go:101]: reverted

File: service/proxy/config_test.go (hunk 11 above)
Comment (askarsabyrov) [reply service/proxy/config_test.go]: reverted, sorry

File: service/proxy/auth/signature_v2.go (hunk 9 above)
Comment (askarsabyrov) [reply service/proxy/auth/signature_v2.go]: rewrited this func to reuse BucketFromHost, made that func public

**Comment by arttor** [comment]
ci check failed. can you please update your branch? main has upgraded dependencies. you can also run `make test` locally to make sure that ci will pass.

**Comment by askarsabyrov** [comment]
@arttor  Rebased with upstream, but it didn’t resolve the CI govulncheck failure. The new failures were coming from golang.org/x/crypto@v0.50.0 advisories fixed in v0.52.0, so I updated golang.org/x/crypto to v0.52.0 and refreshed related golang.org/x/* dependencies in the root module and tool modules.
After i got:
```
Your code is affected by 2 vulnerabilities from 1 module.
This scan also found 1 vulnerability in packages you import and 7
vulnerabilities in modules you require, but your code doesn't appear to call
these vulnerabilities.
Use '-show verbose' for more details.

>>> All reported vulnerabilities are allowlisted in .govulncheck-ignore.yaml — PASS
```

**Review by arttor — APPROVED** [review]
